jsconsole on windows

debris opened the pull request:
changes
- extdep downloads `v8` version `3.15.9`
- findv8 properly finds `v8` on windows
- `readline` is optional for `v8`

arkpar commented on cmake/Findv8.cmake:
```diff
@@ -31,12 +31,32 @@ set(V8_LIBRARIES ${V8_LIBRARY})
 	)
 	
-	set(V8_LIBRARIES optimized ${V8_LIBRARIES} debug ${V8_LIBRARY_DEBUG})
+	set(V8_LIBRARIES ${V8_LIBRARY} ${V8_NO_SNAPSHOT_LIBRARY})
```
also need Winmm.lib here

debris replied:
done

arkpar commented:
JSV8Engine::eval and JSV8Printer::prettyPrint must share the same HandleScope, otherwise it crashes. Remove HandleScope from these functions and place it in JSConsole.h before these functions are called.

arkpar commented:
windows console does not show colored prettyPrint properly
> 2 + 3
←[31m5←[0m

arkpar commented:
There is a crash on closing the window
```
 	ethconsole.exe!v8::internal::OS::DebugBreak(void)	Unknown
 	ethconsole.exe!v8::internal::OS::Abort(void)	Unknown
 	ethconsole.exe!V8_Fatal()	Unknown
 	ethconsole.exe!v8::internal::Isolate::Current(void)	Unknown
 	ethconsole.exe!V8_Fatal()	Unknown
 	ethconsole.exe!v8::internal::Isolate::Current(void)	Unknown
 	ethconsole.exe!V8_Fatal()	Unknown
 	ethconsole.exe!v8::internal::Isolate::Current(void)	Unknown
 	ethconsole.exe!v8::V8::Dispose(void)	Unknown
>	ethconsole.exe!dev::eth::JSV8Env::~JSV8Env() Line 102	C++
 	ethconsole.exe!`dev::eth::JSV8Env::getInstance'::`2'::`dynamic atexit destructor for 's_env''()	C++
 	msvcr120d.dll!_unlockexit()	Unknown
 	msvcr120d.dll!_cexit()	Unknown
 	msvcr120d.dll!_CrtEndBoot()	Unknown

```

debris commented:
@arkpar thanks for tip with handleScopes. I just removed them from JSV8Engine && JSV8Printer and I'm using one created in JSV8Scope.

> windows console does not show colored prettyPrint properly

I removed colored output on windows.

> There is a crash on closing the window

Fixed

arkpar commented on libjsengine/JSV8Engine.cpp:
```diff
@@ -23,6 +23,10 @@
 #include "libjsengine/JSEngineResources.hpp"
+#include "BuildInfo.h"
+
+#define TO_STRING_HELPER(s) #s
```
use DEV_QUOTED_HELPER from libdevcore/Common.h

debris replied:
I do not want to make libjsengine dependent on libdevcore. On macos `libjsengine` must be compiled with old `libstdc++` instead of `libc++`. I do not want to bring more complexity here.

arkpar commented on libjsconsole/JSConsole.h:
```diff
@@ -42,18 +44,35 @@ class JSConsole
-		g_logPost = [](std::string const& a, char const*) { std::cout << "\r           \r" << a << std::endl << std::flush; rl_forced_update_display(); };
+		g_logPost = [](std::string const& a, char const*)
+		{
+		    std::cout << "\r           \r" << a << std::endl << std::flush;
```
Not sure if "\r" works the same way on all platforms.

debris replied:
according to [this stackoverflow answer](http://stackoverflow.com/questions/7013034/does-windows-carriage-return-r-n-consist-of-two-characters-or-one-character) it is

arkpar commented on libjsconsole/JSConsole.h:
```diff
@@ -42,18 +44,35 @@ class JSConsole
-				cmd += std::string(buff);
-				cmd += " ";
+				rl = std::string(buff);
 				free(buff);
```
free should be called for any non-null pointer

debris replied:
fixed in 5a99bc96b792866115a59bce206ffdb9e6b94b32
